aea/cli/upgrade.py

```diff
@@ -0,0 +1,315 @@
+
+
+class ItemRemoveHelper:
+    """Helper to check dependencies on removing component from agent config."""
+
+    def __init__(self, agent_config: AgentConfig) -> None:
+        """Init helper."""
+        self._agent_config = agent_config
+
+    def get_agent_dependencies(self) -> defaultdict:
+        """
+        Get all reverse dependencices in agent.
+
+        :return:  defaultdict with package id: (type, public_id) and set of component that uses this package
+        """
+        return self.get_item_dependencies(self._agent_config, None)
+
+    @staticmethod
+    def get_item_config(item_type: str, public_id: PublicId) -> PackageConfiguration:
+        """Get item config for item,_type and public_id."""
+        return load_item_config(
+            item_type,
+            package_path=Path("vendor")
+            / public_id.author
+            / f"{item_type}s"
+            / public_id.name,
+        )
+
+    @staticmethod
+    def _get_item_requirements(
+        item: Any,
+    ) -> Generator[Tuple[str, PublicId], None, None]:
+        """
+        List all the requiemenents for item provided.
+
+        :return: generator with package ids: (type, public_id)
+        """
+        item_types = ["connection", "skill", "protocol", "contract"]
+        for item_type in item_types:
+            items = getattr(item, f"{item_type}s", set())
+            for item_public_id in items:
+                yield (item_type, item_public_id)
+
+    def get_item_dependencies(self, item: Any, item_id=None) -> defaultdict:
+        """
+        Get item dependencies.
+
+        :return:  defaultdict with package id: (type, public_id) and set of component that uses this package
+        """
+        result: defaultdict = defaultdict(set)
+
+        for dep_item_type, dep_item_public_id in self._get_item_requirements(item):
+            dep_item_id = (dep_item_type, dep_item_public_id)
+            if item_id is None:
+                _ = result[dep_item_id]  # init default dict value
+            else:
+                result[dep_item_id].add(item_id)
+            dep_item = self.get_item_config(dep_item_type, dep_item_public_id)
+            for item_key, deps in self.get_item_dependencies(
+                dep_item, dep_item_id
+            ).items():
+                result[item_key] = result[item_key].union(deps)
+
+        return result
+
+    def check_remove(
+        self, item_type: str, item_public_id: PublicId
+    ) -> Tuple[set, set, dict]:
+        """
+        Check item can be removed from agent.
+
+        required by - set of components that requires this component
+        can be deleted - set of dependencies used only by component so can be deleted
+        can not be deleted  - dict - keys - packages can not be deleted, keys are set of packages requireed by.
+
+        :return: Tuple[required by, can be deleted, can not be deleted.]
+        """
+        item_key = (item_type, item_public_id)
```

Comment:
use `package_id`

Reply:
item type and public id are needed together, i can not remove a package without package type.

Reply:
sorry, think about public id

Reply:
done